apps/prairielearn/src/lib/editors.ts

```diff
@@ -253,28 +300,30 @@ export abstract class Editor {
 
             // This will both discard the commit we made locally and also pull
             // in any new changes from the remote.
-            await cleanAndResetRepository(this.course, gitEnv, job);
+            await cleanAndResetRepository(this.course, `origin/${this.course.branch}`, gitEnv, job);
 
             await writeAndCommitChanges();
 
-            job.info('Push changes to remote git repository');
-            await job.exec('git', ['push'], {
-              cwd: this.course.path,
-              env: gitEnv,
-            });
-            job.data.saveSucceeded = true;
+            try {
+              job.info('Push changes to remote git repository');
+              await job.exec('git', ['push'], {
+                cwd: this.course.path,
+                env: gitEnv,
+              });
+              job.data.saveSucceeded = true;
+            } finally {
+              // Clean up to remove any empty directories that might have been
+              // left behind by operations like renames.
```

Comment:
I'm erring on the side of lots of comments.

```suggestion
              // Clean up to remove any empty directories that might have been
              // left behind by operations like renames. This will also ensure
              // that we get back to a good state if the changes couldn't be
              // pushed to the remote.
```